meeting notes 2021-09-08 (#256)
solid/authorization-panel: meetings/2021-09-08.md

# 2021-09-08 Solid Authorization

https://meet.jit.si/solid-authorization

## Agenda

* Previous meeting minutes https://github.com/solid/authorization-panel/pull/254
* **no access** by default for apps (clients)
  * which use case is this related to?
* [Access Modes](https://github.com/solid/authorization-panel/issues/253)

## Present

* Justin
* Pavlik
* Aaron
* Henry
* Sarven

## Minutes

### Previous meeting minutes

> Henry: I made PR but there were some unclear and missing parts. I asked few questions inline.

Reviewer: ```suggestion
**Henry** I made PR but there were some unclear and missing parts. I asked few questions inline.
```

> **Sarven** I can add commit to that PR

Reviewer 2: ```suggestion

**Sarven** I can add commit to that PR
```

### No access by default for apps

> **PAVLIK** the perception that apps always have all of the access a given authenticated end-user has, but using client-id it is possible to restrict access except for allowed clients. Should the default be to block any clients but permitted ones. Related conversation https://github.com/solid/data-interoperability-panel/issues/139#issuecomment-915102197

Reviewer 2: ```suggestion
**PAVLIK** the perception that apps always have all of the access a given authenticated end-user has, but using client-id, it is possible to restrict access except for allowed clients. Should the default be to block any clients but permitted ones? [Related conversation](https://github.com/solid/data-interoperability-panel/issues/139#issuecomment-915102197).
```

**Sarven** Maybe we should create issue first before we discuss things here.
There is a discussion in WAC [issue 81: Client identification](https://github.com/solid/web-access-control-spec/issues/81) 

> **Henry** I think there is a thread on this in our discussion channel: on [proofs and Origins](
> https//github.com/solid/authorization-panel/discussions/223)

Reviewer 2: ```suggestion
**Henry** I think there is a thread on this in our discussion channel: on 
[proofs and Origins](https//github.com/solid/authorization-panel/discussions/223)
```

>  One thing that came out of that discussion, is that I don't think that this problem applies to [HTTPSig](https://github.com/solid/authentication-panel/blob/main/proposals/HttpSignature.md).  There a wallet is signing every request with the key of the user. The wallet, can thus be thought to be authorizing access to the resource independent of the client app, since it is signing a specific request. The wallet is responsible for enforcing user authorization on the client, by choosing which resources it will sign requrests for. 

Reviewer 2: ```suggestion
One thing that came out of that discussion, is that I don't think that this problem applies to [HTTPSig](https://github.com/solid/authentication-panel/blob/main/proposals/HttpSignature.md).  There, a wallet is signing every request with the key of the user. The wallet can thus be thought to be authorizing access to the resource independent of the client app, since it is signing a specific request. The wallet is responsible for enforcing user authorization on the client, by choosing which resources it will sign requests for. 
```

> **Justin** We have 2 dimensions to consider: a) the social entity (eg. me) and b) the application used (by me). Should we discuss it here? Do I want to restrict access to only a subset of application which I use?

Reviewer 2: ```suggestion
**Justin** We have 2 dimensions to consider: a) the social entity (e.g., me) and b) the application used (by the social entity, e.g., me). Should we discuss it here? Do I want to restrict access to only a subset of applications which I use?
```

I would say yes, there is a benefit.

> Already work on Solid-OIDC provides us with `client_id`. HTTPSig can deal with this issue in a different way.

Reviewer 2: ```suggestion
Already, work on Solid-OIDC provides us with `client_id`. HTTPSig can deal with this issue in a different way.
```

There seems to be notion that as soon as app gets a token, it can access anything that user can access.
We don't want to create such perception. As soon as we review draft for ACP or WAC+ or something else.
We will evaluate it on how it gives restrictions on clients.

> **Henry** Note with HttpSig the Wallet can have policies about which apps it gives signtatures to for what sets of resources, so that access control does not need to be done in server ACRs.

Reviewer 2: ```suggestion
**Henry** Note that with HttpSig, the Wallet can have policies about which apps it gives signatures to for what sets of resources, so that access control does not need to be done in server ACRs.
```

There seems to be a requirement for Solid-OIDC that servers need to be authorizing clients.
This is not needed for HTTPSig. 
I am ok to support the Solid OIDC use case.

**Aaron** I see authorizing user vs clients as different dimensions. One would be authorizing users
/ social agents. The other one is authorizing clients. As soon as authorization system can

> identify both it can do its job. I know that ACP allows that.

Reviewer 2: ```suggestion
identify both, it can do its job. I know that ACP allows that.
```

> **Henry**: A so if we are speaking of setting access control rules that the Wallet agent can use before signing HTTP headers, I think we should be able to use the WAC ontology to do that. The Wallet needs to know which set of resoures a client can access in which modes before signing the headers. 

Reviewer 2: ```suggestion
**Henry**: So if we are speaking of setting access control rules that the Wallet agent can use before signing HTTP headers, I think we should be able to use the WAC ontology to do that. The Wallet needs to know which set of resources a client can access in which modes, before it can sign the headers. 
```

**Pavlik**  Ah perhaps you can write that up as an issue Henry.

I would like to propose using GNAP terminology of End-user, client (instance), Resource Owner.

### [Access Modes](https://github.com/solid/authorization-panel/issues/253)

**Justin** https://solid.github.io/authorization-panel/authorization-ucr/#collection-readcreatedelete

> **Justin** We have modes which are overloaded. Given write one can delete, create, update.

Reviewer 2: ```suggestion
**Justin** We have modes which are overloaded. Given `write`, one can `delete`, `create`, `update`.
```

This leads to over permissioning agents. I think this should be more urgent issue to resolve.
I think aligning it around CRUD is not a radical proposition. Kjetilk has done really good job
describing the issue. I recall Matthieu suggesting to put it in separate document.

**Pavlik** I'm not sure if protocol spec has to define both operations and access modes. Also how do we imagine mapping between those two?

**Henry** 

>  It is a great discussion. But it is branching a lot. Whould we move it over to the discussions area, which also has threading?

Reviewer 2: ```suggestion
 It is a great discussion. But it is branching a lot. Should we move it over to the discussions area, which also has threading?
```

>  I am trying to find mathematical backing for the decisions we could make, so that we don't have to jump back and forward all the time. I think we have clearly Read/Write (those come from Lenses) and there could be subsets under Write for Create Update Delete. 
> Delete is an interesting question: does it need permissions on the container?

Reviewer 2: ```suggestion
 I am trying to find mathematical backing for the decisions we could make, so that we don't have to jump back and forward all the time. I think we have clearly `Read`/`Write` (those come from Lenses) and there could be subsets under `Write` for `Create` `Update` `Delete`. 
`Delete` is an interesting question: does it need permissions on the container?
```

**Justin** I don't forsee coming back 6 months later or 6 years later and saying - i need another mode.
There has been some arguments that we will add some later, I think that our current set needs improvement
but after that they should possibly last forever.

**Barath** I agree, that well chosen modes will last for long time. Even substandard modes serve in existing systems.

> **Justin** I think delete is good example, it can be set on the level of container and on individual resources. I don't see argument that there could be delete without implicit update of containment triples.

Reviewer 2: ```suggestion
**Justin** I think `delete` is good example; it can be set on the level of container and on individual resources. I don't see argument that there could be `delete` without implicit update of containment triples.
```

Argument that something else gets changed doesn't sound for me like recipe for good authorization system.

> **Pavlik** I think create and delete without allowing updating client managed triples in the

Reviewer 2: ```suggestion
**Pavlik** I think `create` and `delete` without allowing updating client managed triples in the
```

container is a good scenario that is needed to be addressed.

**Sarven** If you look at access modes in a vacum there may be all kind of modes that can be discussed.
If we have protocol which has hierarchical containment ...

> Protocol usually starts with read and write. When you interacting - making a request against specific
> type of resource eg. container. It is processing operation, where protocol defines what happens.
> In WAC in order to delete something, one needs to have write on the resource and write on container

Reviewer 2: ```suggestion
Protocol usually starts with `read` and `write`. When you're interacting — making a request against specific
type of resource, e.g., container. It is processing operation, where protocol defines what happens.
In WAC, in order to delete something, one needs to have `write` on the resource and `write` on container
```

containing it.

Another thing that seems to be conflated is resource level and content level.

[isssue 121: Towards an authorization framework for the Solid ecosystem](https://github.com/solid/authorization-panel/issues/121)

> We still haven't solved that issue, it can help us clarify where each use case fits in.

Reviewer 2: ```suggestion
We still haven't solved that issue; it can help us clarify where each use case fits in.
```

I think we are ripping this issue apart instead of tackling it directly.
I see operations in the protocol territory. I don't think solution needs to demand some particular
access modes.

> If we look at PUT, from HTTP perspective you either create issue first or in subsequent requests
> you replace / update it.

Reviewer 2: ```suggestion
If we look at `PUT`, from HTTP perspective you either `create` issue first, or, in subsequent requests,
you `replace` / `update` it.
```

> **Henry** I've been looking for some math that could help put some clarity in the problem. We are dealing with a prtocol between a client and a server, so that for example limits what we need to cover: we don't need to cover that the server has writes to make the changes.
> We sould assume that server has rights to do what it needs to do. I was mentioning Lenses: these are a generalisation of what we know in Java Beans with getters and setters. Having getters and setters one can do CRUD.

Reviewer 2: ```suggestion
**Henry** I've been looking for some math that could help put some clarity in the problem. We are dealing with a protocol between a client and a server, so that, for example, limits what we need to cover: we don't need to cover that the server has rights to make the changes.
We should assume that server has rights to do what it needs to do. I was mentioning Lenses: these are a generalisation of what we know in Java Beans with getters and setters. Having getters and setters, one can do CRUD.
```

I'm looking for papers on those topics. One thing that stands out clearly is that `Control` is not part of the space. It seems to be a
different concept. We could remove control and focus how CRUD can be modeled.

> I'm trying to understand why one would need access to the container. I don't think one needs write
> on the container to delete resource. We can see if there are any potential security holes by building up an example with and without write. 

Reviewer 2: ```suggestion
I'm trying to understand why one would need access to the container. I don't think one needs `write`
on the container to `delete` resource. We can see if there are any potential security holes by building up an example with and without `write`. 
```

> **Justin** I'm able to delete comment in a container of comments. I should be able to delete mine but
> not anyone's else. Delete on the container, meaning that it applies to all the comment would allow
> deleting all the comments.

Reviewer 2: ```suggestion
**Justin** I'm able to `delete` a comment in a container of comments. I should be able to delete mine but
not anyone's else. `Delete` on the container, meaning that it applies to all the comments would allow
deleting all the comments.
```

> **Sarven** Delete on the container doesn't imply delete on all the contained resources.

Reviewer 2: ```suggestion
**Sarven** `Delete` on the container doesn't imply `delete` on all the contained resources.
```

> **Henry** We can take your example and do our evaluation process on it. For example Sarven

Reviewer 2: ```suggestion
**Henry** We can take your example and do our evaluation process on it. For example, Sarven
```

could should how WAC can do it.

> **Sarven** Only the agent who has write on the resource can delete this resource.
> 
> **Justin** Write includes delete but it gives another access as well.

Reviewer 2: ```suggestion
**Sarven** Only the agent who has `write` on the resource can delete this resource.

**Justin** `Write` includes `delete`, but it gives another access as well.
```

## Actions

> **CSarven** Fix PR of previous meeting
> 
> **Pavlik** create issue proposing GNAP terminology
> 
> **Henry** Propose using WAC to limit wallets to sign requests for apps.
> 
> **Justin**: write up comparison for a use case on modes

Reviewer 2: ```suggestion
* **CSarven** Fix PR of previous meeting

* **Pavlik** create issue proposing GNAP terminology

* **Henry** Propose using WAC to limit wallets to sign requests for apps.

* **Justin**: write up comparison for a use case on modes
```
